### ISSUE TITLE
IP address in the auth

### ISSUE THREAD

**Issue by dufourc** [emqx/emqx-auth-mysql#243]
Hey, 

I see in the documentation that IP address is only usable for ACL query and not inside auth query. I've tried and this is actually not working, I received:

```
select password from mqtt_user where username = 'stats_pollers' and ip = '\'%a\'' limit 1;
```

if I have 

```
auth.mysql.auth_query = select password from mqtt_user where username = '%u' and ip = '%a' limit 1
``` 

in my emqx_auth_mysql.conf. 

When I look at the code, I can see that in emqx_auth_mysql.erl:check fun and semqx_acl_mysql.erl:do_check_acl are both using emqx_auth_mysql_cli:query(Pool, AclSql, AclParams, ClientInfo) which is using the replvar:
https://github.com/emqx/emqx-auth-mysql/blob/master/src/emqx_auth_mysql_cli.erl#L77.

Why is it working with one and not the other?
Is the IpAddr not accesible while authenticating? But in this case, why it is accessible in the auth.http.auth_req.params from https://github.com/emqx/emqx-auth-http?

**Comment by HJianBo** [comment]
Sorry, @dufourc, this is an imperfection in the documentation, we will fix it.

These placeholders' support is inconsistent in each plugin. My suggestion is to try to use the ones mentioned in the configuration file comments. (Although it is sometimes not accurate...)

Of course, if you can read the code, this place is the most accurate.

I'm very sorry for the confusion these inconsistent implementations have brought you. We've planned to unify them in a subsequent release!

**Comment by dufourc** [comment]
Don't worry, it's hard to keep a documentation up to date for such a big product! 

I would like to use the IP in the authentication query, do you have any ideas why it is not working?

**Comment by HJianBo** [comment]
It's a bug actually. The `emqx_auth_mysql_cli:replvar/2` should get the '%a' field by `peerhost` instead of `peername`.

We will fix it and plan to release it in 4.2.2. Thanks @dufourc :)

**Comment by HJianBo** [comment]
In addition, I'm not quite sure why the client's IP address needs to be used for authentication instead of the username/client ID?

**Comment by dufourc** [comment]
You're right, for authentication username/password or cookie should be enough. But we only use IP in ACLs so to make sure, the user we authenticate has the correct permission, we need to check that the IP address matches the username/password.

In our project, we have two kind of EMQX client: 

- pollers: extract information from a machine (server, switch, DB etc..) then publish to EMQX.
- browsers: subscribe to it to have some graphics/tables about these devices.

In the ACL file, we only use IP rules. Indeed browsers appears to come all from the same address (Nginx proxy) and pollers are locals. 
To make the ACL rules relevant, we need to check the IP/username/password during the authentication.

Another possibility would have been to use a plugin allowing more complex ACLs like SQL or HTTP, but it would have been more work, more  queries/requests made and the cache to handle.

We are as well checking the clientID (that need to contain a unique prefix stored in our session), to kickout users if needed. 

It might be wrong, but that's the solution we've come up with

**Comment by dufourc** [comment]
> It's a bug actually. The `emqx_auth_mysql_cli:replvar/2` should get the '%a' field by `peerhost` instead of `peername`.
> 
> We will fix it and plan to release it in 4.2.2. Thanks @dufourc :)

Thanks, amazing!

**Comment by dufourc** [comment]
@HJianBo I've seen the change in the code and this is working for me. Thank you very much :) I'm closing the issue
